### PR TITLE
Centralize unparsing of comment references

### PR DESCRIPTION
**Pull request by jcollins-g**
Before implementing new parsing of comment references, we should clean up how dartdoc gets comment references in the first place.  Unparsing them is now centralized which should make it more clear how the analyzer AST could help dartdoc by providing the raw string.

Breaking change:  commentRefs, a member of ModelElements, is a map now instead of a list.

### CODE REVIEW

File: lib/src/comment_references/model_comment_reference.dart [hunk 1]
```diff
@@ -0,0 +1,29 @@
+// Copyright (c) 2020, the Dart project authors. Please see the AUTHORS file
```
Comment (srawlins) [inline lib/src/comment_references/model_comment_reference.dart:1]: 2021?

File: lib/src/markdown_processor.dart [hunk 2]
```diff
@@ -126,10 +126,6 @@ final RegExp nonHTML =
 /// parentheses.
 final _trailingIgnorePattern = RegExp(r'(<.*>|\(.*\)|\?|!)$');
 
-@Deprecated('Public variable intended to be private; will be removed as early '
```
Comment (srawlins) [inline lib/src/markdown_processor.dart:129]: Woohoo!

File: lib/src/markdown_processor.dart [hunk 3]
```diff
@@ -369,26 +357,25 @@ MatchingLinkResult _getMatchingLinkElementLegacy(String codeRef,
 }
 
 /// Given a set of commentRefs, return the one whose name matches the codeRef.
```
Comment (srawlins) [inline lib/src/markdown_processor.dart:359]: Looks like the doc comment doesn't apply any more.

File: lib/src/comment_references/model_comment_reference.dart [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+  ModelCommentReference(CommentReference ref, ResourceProvider resourceProvider)
+      : codeRef = _referenceText(ref, resourceProvider),
+        staticElement = ref.identifier.staticElement;
+
+  /// "Unparse" the code reference into the raw text associated with the
+  /// [CommentReference].
+  static String _referenceText(
+      CommentReference ref, ResourceProvider resourceProvider) {
+    var contents = getFileContentsFor(
```
Comment (srawlins) [inline lib/src/comment_references/model_comment_reference.dart:25]: Is this super expensive?

**Review by srawlins — COMMENTED** [review]
Sorry, I know just a draft, but I couldn't help but look!

File: lib/src/comment_references/model_comment_reference.dart (hunk 4 above)
Comment (jcollins-g) [reply lib/src/comment_references/model_comment_reference.dart:25]: Fortunately not.  It is cached so we only pay a read penalty once per file.

File: lib/src/comment_references/model_comment_reference.dart (hunk 1 above)
Comment (jcollins-g) [inline lib/src/comment_references/model_comment_reference.dart]: indeed

File: lib/src/markdown_processor.dart (hunk 3 above)
Comment (jcollins-g) [reply lib/src/markdown_processor.dart]: Wrote a more complete doc.

**Review by srawlins — APPROVED** [review]

**Comment by jcollins-g** [comment]
I have redone some benchmarks before marking this as ready for review, to make sure the changes to comment reference handling didn't harm performance.  Fortunately, the delta is negligible.

**Pull request merged**